packages/aws-cdk-lib/aws-dynamodb/lib/table.ts

```diff
@@ -564,7 +564,7 @@ export abstract class TableBase extends Resource implements ITable, iam.IResourc
    * @param actions The set of actions to allow (i.e. "dynamodb:PutItem", "dynamodb:GetItem", ...)
    */
   public grant(grantee: iam.IGrantable, ...actions: string[]): iam.Grant {
-    return iam.Grant.addToPrincipalOrResource({
+    return iam.Grant.addToPrincipal({
```

Comment:
Why does this need to become `addToPrincipal`? We are for example already using the resource policy to grant cross-account access to a table directly. We build the policy in CDK from external account `ArnPrincipal`s.

Reply:
Because addToPrincipalOrResource never adds to the resource, as IAM is missing methods. So I'm putting it back to how it was before I changed it until I have time to implement the missing methods in IAM

Reply:
I don't fully understand what methods are missing. Looking at `Grant`, it seems to be handling this already, if given the `resource` in the call: https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-iam/lib/grant.ts#L145-L157. It's just that `Table.addToResourcePolicy` is not functional, but instead adds statements to a policy instance not bound with the internal `CfnTable`.

Reply:
@LeeroyHannigan I want to check if I follow your line of thoughts here.

I see the `iam.Grant.addToPrincipalOrResource` is implemented [here](https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-iam/lib/grant.ts#L122). It basically does 2 things, call `addToPrincipal` and a bit further down on line 151, it calls `options.resource.addToResourcePolicy`, which means the `Table.addToResourcePolicy`. Since the method is reportedly a no-op (from https://github.com/aws/aws-cdk/issues/30793), you are saying your change here from `addToPrincipalOrResource` to `addToPrincipal` will not change any actual behaviour.

I agree with that. But if I understand @everilae correctly, even though `Table.addToResourcePolicy` was a no-op, there may be CDK users who are currently relying on what `iam.Grant.addToPrincipalOrResource` returns (it returns the resource policy statement whereas `iam.Grant.addToPrincipal` returns principal policy statement.). This allows CDK users to use this method to build the resource policy instead of creating their own.

Reply:
Hi,

We rely on the "...OrResource" part, or would like to, once all this is fixed. If the call is replaced with just "addToPrincipal", then attempts at granting to external accounts would not work.